service/db/boltdb.go

```diff
@@ -46,3 +47,37 @@ func DB() *bbolt.DB {
 	once.Do(initDB)
 	return db
 }
+
+// The function should return a dirty flag.
+// If the dirty flag is true and there is no error then the transaction is commited.
+// Otherwise, the transaction is rolled back.
+func Transaction(db *bbolt.DB, fn func(*bbolt.Tx) (bool, error)) error {
+	tx, err := db.Begin(true)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
```

Comment:
Twice possible and unexpected Rollback?
And unpexpected Rollback after Commit?

Reply:
It's OK. I copied the code from boltdb.

```go
func (db *DB) Update(fn func(*Tx) error) error {
	t, err := db.Begin(true)
	if err != nil {
		return err
	}

	// Make sure the transaction rolls back in the event of a panic.
	defer func() {
		if t.db != nil {
			t.rollback()
		}
	}()

	// Mark as a managed tx so that the inner function cannot manually commit.
	t.managed = true

	// If an error is returned from the function then rollback and return error.
	err = fn(t)
	t.managed = false
	if err != nil {
		_ = t.Rollback()
		return err
	}

	return t.Commit()
}
```

Only the first commit or rollback will work.

Reply:
Fixed

Reply:
Thanks, it looks good!
Did you re-test the disk writes?

Reply:
Thanks, it looks good!
Did you re-test the disk writes?

Reply:
The disk throughput drops from 17KB/s to almost 0KB/s. 
I will run the program for a week to collect long-time running data.

Reply:
Great!